File: pdam-frontend/pdam_app/lib/models/post/GetPostDto.dart
import '../user/GetProfile.dart';

class GetPostDto {
  late int id;
  late String? affair;
  late String content;
  late List<String> imgPath;
  late UserWhoPost userWhoPost;
  late int usersWhoLiked;
  late int comments;
  late bool likedByUser;
  late String postDate;

  GetPostDto(
      {required this.id,
      this.affair,
      required this.content,
      required this.imgPath,
      required this.userWhoPost,
      required this.usersWhoLiked,
      required this.comments,
      required this.likedByUser,
      required this.postDate});

  GetPostDto.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    affair = json['affair'];
    content = json['content'];
    imgPath = json['imgPath'].cast<String>();
    userWhoPost = (json['userWhoPost'] != null
        ? new UserWhoPost.fromJson(json['userWhoPost'])
        : null)!;
    usersWhoLiked = json['usersWhoLiked'];
    comments = json['comments'];
    likedByUser = json['likedByUser'];
    postDate = json['postDate'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this.id;
    data['affair'] = this.affair;
    data['content'] = this.content;
    data['imgPath'] = this.imgPath;
    data['userWhoPost'] = this.userWhoPost.toJson();
    data['usersWhoLiked'] = this.usersWhoLiked;
    data['comments'] = this.comments;
    data['likedByUser'] = this.likedByUser;
    data['postDate'] = this.postDate;
    return data;
  }
}

File: pdam-backend/pdam.api/src/main/java/com/salesianostriana/dam/pdam/api/post/dto/GetPostDto.java
package com.salesianostriana.dam.pdam.api.post.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.salesianostriana.dam.pdam.api.post.model.Post;
import com.salesianostriana.dam.pdam.api.user.dto.UserWhoLikeDto;
import com.salesianostriana.dam.pdam.api.user.model.User;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class GetPostDto {

    private Long id;
    private String affair;
    private String content;
    private List<String> imgPath;
    private UserWhoLikeDto userWhoPost;
    private int usersWhoLiked;
    private int comments;
    private boolean likedByUser;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd/MM/yyyy HH:mm:ss")
    private LocalDateTime postDate;

    public static GetPostDto of(Post post, User user){
        return GetPostDto.builder()
                .id(post.getId())
                .affair(post.getAffair())
                .content(post.getContent())
                .imgPath(post.getImgPaths())
                .userWhoPost(UserWhoLikeDto.of(post.getUserWhoPost(), user))
                .usersWhoLiked(post.getUsersWhoLiked() == null ? 0 : post.getUsersWhoLiked().size())
                .comments(post.getComments() == null ? 0 : post.getComments().size())
                .postDate(post.getPostDate())
                .likedByUser(user.getLikedPosts().stream().filter(p -> Objects.equals(p.getId(), post.getId())).toList().size() > 0)
                .build();
    }

}
